### DISCUSSION TITLE
Multiple PN532 card readers on ESP32

### DISCUSSION THREAD

**Discussion by belveder79** [arendst/Tasmota#20849] [Q&A]
I'd need to connect a total of 8 PN532 card readers to a single ESP32. It seems that the driver for the PN532 is relatively straight-forward; it is basically a serial. There should be lots of pins available, but I wonder how to best "duplicate" the driver to make this possible... any hints?

**Comment by sfromis** [comment]
If you duplicate driver (or extend it to support multiple device), you'd run into the issue of ESP32 only having a low number of serial ports available, meaning that 8 ports is not supported.

**Reply by barbudor** [reply]
Not sure how the PN532 protocol is but if it only respond to request, it could be possible using a bit of external logic (demultiplexer) to make some round mecanism.

If it is supposed to send a message on its own such like when a card is presented, then that would not be possible. So limited to 3 per ESP.

**Comment by belveder79** [comment]
The pn532 driver polls every 250ms. In my use case I would have the tags
put there and stay there. It does not matter if a scanning round of eg. 10
readers would take 10 secs or more, so i would just multiplex the tx/rx
lines, right?

Barbudor ***@***.***> schrieb am Sa. 2. März 2024 um 19:33:

**Reply by barbudor** [reply]
Yes
Now the question would be how to do that nicely both in hardware and driver
What I would do is
- Hardware wise, 
  - all serial from PN532 to ESP32.RX can be multiplexed using simple diodes (K on PN side, A on ESP side). 
  - use a 74xx138 with A0..A2 connected to 3 GPIO, /G0=ESP.TX, /G1=GND, G2=VCC and Y0..Y7 to the PN.TX
- Software wise you would need 3 GPIO PN532_SEL0..PN532_SEL1 and use that to drive the round-robin
To avoid breaking change, if the SEL GPIO are not define, the driver should remains in single-device mode

You can also use a 74xx151 as a multiplexer for PN.TX to ESP.RX

The techno `xx` need to be carefully chosen depending if the PN532 is a 3V3 or 5V device

**Reply by sfromis** [reply]
As the driver has a single data structure with various device status data, including some which looks like being for more than a single poll operation, like not reading more than once in 2 seconds, I'd expect a risk of the driver getting confused if you multiplex serial to multiple device.

However, as I've not analyzed details of driver operation, I'm not trying to make an assertion either way; merely suggesting that it may not be a safe assumption that it would work without a hitch.

**Reply by barbudor** [reply]
I didn't said that the data structure shouldn't also be adapted like everytime you improve a driver from single-device to multiple-device
I was just addressing how to multiple serial links if the driver can work is a pure request-response polling mode.

**Reply by belveder79** [reply]
Well, there is an underlying more philosophical question if the way e.g. the I2C is replicated for 2 I2C busses in tasmota for esp32 is also done the "nice" way... I just looked into this and it is merely a new variable that the methods apply to, but also with #defines and if clauses,  also not scalable to 4 I2C busses, for example.

I did not do too much of C programming in the tasmota framework yet and that change seems to be a more radical one. I already thought to just replicate the driver in parts in berry, but I'm not sure about that either...

**Reply by barbudor** [reply]
I2c is fully handled by software but-banging so there is no real difficulty to multiply them
Serial is by default using UARTs which are limited to 3 on the esp32
It is theorically possible to use software Serial but  while there is some interrest on esp8266 which as only 1 UART, it was decided to to port software serial as 3 UARTs (or 2 on C3) was considered enough.

Without more UARTs, even in Berry you will need some external multiplexing.
But I can understand it could be easier to do it in Berry than diving into Tasmota code

**Reply by sfromis** [reply]
Simplest case for multiple devices on same bus (like many I2C sensors) is when each device has some sort of id, which you can poll, without worrying about pins per device. But still the driver needs to support multiples, meaning that it is still not like "so i would just multiplex".

This is basically how Modbus devices can coexist on the same serial bus.

**Comment by belveder79** [comment]
So I tried to just switch the tx/rx pins of 2 devices during runtime and I
think the idea of just using a multiplexer should work as both were reading
correctly. The only thing missing is the wakeup command that needs to be
sent to each reader after applying power, so only one works initially, and
if you connect the second and reboot, the second starts working and so on…

 I think sending this sequence just to all readers one after the other in a
berry script right at the start should make all work and that would do the
trick… it is not a super sexy solution, but way easier than changing the
driver… will report…

sfromis ***@***.***> schrieb am Sa. 2. März 2024 um 22:04:

**Reply by Jason2866** [reply]
Could imagine you run in some UART hardware serial problems. Since there is a (hardware) buffer you have to take care to read and flush serial the "right" way.

**Reply by belveder79** [reply]
yes, that might indeed happen. Let's see what comes out once I have the multiplexer
